### PR TITLE
Tensor type initcontext

### PR DESCRIPTION
**Pull request by jijoongmoon**
## Commits to be reviewd in this PR

<details><summary>   [ Mixed Tensor ] add tensor type property in initContex  </summary><br />

This PR add the tensor type (Format, Weight Tensor DataType,
Activation Tensor DataType) in initContext.
- Remove the tensor type variables and setter, getter member function
in layer, layer_devel, loss layer etc.
- add tensor type setter in initContext
- set the var_grad ( input & ouput ) Tensor Type according to model
Tensor Data Type.
- Add ModelTensorTypeInfo : eg. FP16_FP16 ( Weight FP16, Activation
FP16 )

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by:jijoong.moon <jijoong.moon@samsung.com>
</details>

### CODE REVIEW

File: nntrainer/layers/layer_context.h [hunk 1]
```diff
@@ -51,7 +51,38 @@ class InitLayerContext {
   InitLayerContext(const std::vector<TensorDim> &dim,
                    const std::vector<bool> &req_out_connected, bool in_place_,
                    const std::string &n = "", const std::string &prefix_ = "",
-                   const float max_norm = 0.0);
+                   const float max_norm = 0.0,
+                   std::array<const std::string, 3> tensor_type_ = {
+                     "NCHW", "FP32", "FP32"});
```
Comment (skykongkong8) [inline nntrainer/layers/layer_context.h:56]: Why does this holds 3 string components? Is it for upcoming mixed-precision update?

File: nntrainer/tensor/tensor.cpp [hunk 2]
```diff
@@ -166,17 +166,17 @@ void Tensor::allocate() {
...
         delete mem_data;
       });
 
     } else if (getDataType() == ml::train::TensorDim::DataType::FP16) {
 #ifdef ENABLE_FP16
-      mem_data = new MemoryData((void *)(new _Float16[dim.getDataLen()]()));
+      mem_data = new MemoryData((void *)(new _Float16[dim.getDataLen()]{}));
       data = std::shared_ptr<MemoryData>(mem_data, [](auto *mem_data) {
-        delete[](_Float16 *) mem_data->getAddr();
+        delete[] mem_data->template getAddr<_Float16>();
```
Comment (skykongkong8) [inline nntrainer/tensor/tensor.cpp:179]: If they are hard-coded with _Float16, will there be updates for selecting either __fp16 and _Float16 under CPU Architecture/compiler?

**Review by baek2sm — APPROVED** [review]

File: nntrainer/layers/layer_context.h (hunk 1 above)
Comment (djeong20) [reply nntrainer/layers/layer_context.h:56]: format, weight, activation

File: nntrainer/layers/layer_node.cpp [hunk 3]
```diff
@@ -525,6 +523,14 @@ InitLayerContext LayerNode::finalize(const std::vector<TensorDim> &input_dims) {
       NNTR_THROW_IF(input_dims != actual_prop_dims, std::invalid_argument)
         << "calculated input dimension is different from given input_shape "
            "property";
+      for (auto d : actual_prop_dims) {
+        d.setDataType(
+          str_converter<enum_class_prop_tag, nntrainer::TensorDataTypeInfo>::
+            from_string(tensor_type[2]));
```
Comment (djeong20) [inline nntrainer/layers/layer_node.cpp:529]: what about tensor_type[1]?

**Review by djeong20 — APPROVED** [review]
LGTM!

**Comment by skykongkong8** [comment]
LGTM!

File: nntrainer/tensor/tensor.cpp (hunk 2 above)
Comment (jijoongmoon) [reply nntrainer/tensor/tensor.cpp:179]: Next PR will be support __fp16 & _Flaot16 using compiler macro. 
which one is better _Half ?? or __FP16??  Let me know.

**Pull request merged**